### PR TITLE
Database upgrade and core enhancements

### PR DESCRIPTION
**Pull request by subhamX**
## Description
The following PR upgrades the database from shelve to MongoDB. It restructures the repository and dockerizes the frontend, backend for both development and production environment. It also includes some new frontend-enhancements.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the *default* branch of the repository (`main` or `master`).
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: backend/scripts/build_db_script.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
+import json
+import datetime
+import pymongo
+import os
+from dotenv import load_dotenv
+
+# Loading env variables file
+load_dotenv('../.env')
```
Comment (soccerdroid) [inline backend/scripts/build_db_script.py]: use the os module to get relative paths too

File: backend/scripts/build_db_script.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import pymongo
+import os
+from dotenv import load_dotenv
+
+# Loading env variables file
+load_dotenv('../.env')
+
+DIST = "D"  # Prefix in every node distance key [eg. D1, D2]
+REV_DIST = "RD"
+INPUT_FILENAME = "fdg_input_file.json"
```
Comment (soccerdroid) [inline backend/scripts/build_db_script.py]: if this file name cannot change, add a comment indicating this

File: backend/src/linked_commons/views.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+from django.http import JsonResponse, FileResponse
+import shelve
+from linked_commons.models import Node
+import random, string
+import json
+import pymongo
+from config.settings import MONGO_DB_CRED
+
+
+LANDING_GRAPH_META = {
```
Comment (soccerdroid) [inline backend/src/linked_commons/views.py:13]: add a comment here so devs understand this setting

File: docker-compose.yml [hunk 4]
```diff
@@ -1,4 +1,46 @@
...
+    build:
+      context: frontend/
+      dockerfile: Dockerfile
+    volumes:
+      - './frontend/:/frontend/'
+      - '/frontend/node_modules'
+    ports:
+      - 3000:3000
+    environment:
+      - CI=true
```
Comment (soccerdroid) [inline docker-compose.yml:40]: Why is this true?

**Review by soccerdroid — CHANGES_REQUESTED** [review]

**Comment by soccerdroid** [comment]
In addition, please add a .env template file, with all the needed environment variables listed here. This is the file that should be loaded later by the apps in order to get the variables values.

File: backend/scripts/build_db_script.py (hunk 2 above)
Comment (subhamX) [reply backend/scripts/build_db_script.py]: No, the user can change the filename. Maybe I should rename it to `INPUT_FILE_PATH` and add a comment for better clarity.

File: docker-compose.yml (hunk 4 above)
Comment (subhamX) [reply docker-compose.yml:40]: There are some issues with `react-scripts`, and even after using `tty: true` on Linux machines, the frontend build was failing. Although, on my `wsl` it was working fine. I have downgraded the react-scripts version and set `CI=true` as one of the users suggested that using it fixes the issue[[Link](https://github.com/facebook/create-react-app/issues/8688#issuecomment-602084087)]. I assumed that it might be a solution for Linux machines. 

@mathemancer could you please confirm if you are able to set up the development version or not.

**Comment by subhamX** [comment]
> In addition, please add a .env template file, with all the needed environment variables listed here. This is the file that should be loaded later by the apps in order to get the variables values.

There is already a `.env` file in `backend/` with some dummy values. I have added comments for more clarity.

File: backend/Dockerfile.prod [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+# install dependencies
+RUN pip install --upgrade pip
+COPY ./requirements.txt .
+RUN pip install -r ./requirements.txt
+
+# copy project
+COPY ./src ./src
+COPY ./nginx ./nginx
+COPY ./.env ./
```
Comment (mathemancer) [inline backend/Dockerfile.prod]: I don't think you need to copy `./nginx` or `./.env` into the image.

File: backend/nginx/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,4 @@
+FROM nginx:1.19.0-alpine
+
+RUN rm -f /etc/nginx/conf.d/default.conf
+COPY nginx.conf /etc/nginx/conf.d
```
Comment (mathemancer) [inline backend/nginx/Dockerfile]: Is this file still needed?

File: backend/nginx/nginx.conf [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+server {
+    listen 80;
+    location / {
+        proxy_pass http://linked_commons;
+        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+        proxy_set_header Host $host;
+        proxy_redirect off;
+    }
+}
```
Comment (mathemancer) [inline backend/nginx/nginx.conf]: Is this still needed?

File: backend/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,6 @@
+Django==2.2.13
+django-cors-headers
+pymongo
+dnspython
+gunicorn
+python-dotenv
```
Comment (mathemancer) [inline backend/requirements.txt]: Please sort these and add specific versions (this will be better for the prod build.  Alternatively, freeze versions in a separate `requirements.prod.txt` (or the like)

File: backend/scripts/build_db_script.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+try:
+    USERNAME = os.environ['MONGO_INITDB_ROOT_USERNAME']
+    PASSWORD = os.environ['MONGO_INITDB_ROOT_PASSWORD']
+    MONGO_DB_NAME = os.environ['MONGO_DB_NAME']
+    MONGO_COLLECTION_NAME = os.environ['MONGO_COLLECTION_NAME']
+except KeyError as e:
+    raise Exception(f"Undefined ENV variable {e.args[0]}")
+
+# Change this to container_name:port inside docker container
+HOSTNAME = "localhost:27017"
```
Comment (mathemancer) [inline backend/scripts/build_db_script.py]: Please make the `host` part of this either an argument or an environment variable.

File: frontend/.env [hunk 10]
```diff
@@ -0,0 +1,2 @@
+REACT_APP_SERVER_BASE_ENDPOINT_DEV=http://localhost:8000/api
+REACT_APP_SERVER_BASE_ENDPOINT_PROD=http://localhost:8000/api
```
Comment (mathemancer) [inline frontend/.env]:
```suggestion
REACT_APP_SERVER_BASE_ENDPOINT_PROD=http://dataviz.creativecommons.engineering/api
```

This needs to be the real value to automate the frontend build.

**Review by mathemancer — CHANGES_REQUESTED** [review]
I made a few specific change requests.  In general, try to avoid the message:
```
\ No newline at end of file
```
from `git`.  (you can also see little red icons in the code on GH where this is an issue.)

Overall, great work!

File: docker-compose.yml (hunk 4 above)
Comment (soccerdroid) [reply docker-compose.yml:40]: Ok, what about downgrading to the 3.4.0 solution mentioned on the thread? It's just weird to see that flag enabled when we still don't have continuous integration testing

File: backend/.env [hunk 11]
```diff
@@ -0,0 +1,28 @@
+# Mongo DB Configuration
```
Comment (soccerdroid) [inline backend/.env:1]: Can we add the variable REACT_APP_SERVER_BASE_ENDPOINT, and keep this single file for all the environment?

**Review by soccerdroid — CHANGES_REQUESTED** [review]

**Comment by soccerdroid** [comment]
Just a question, are the .conf files needed?

File: docker-compose.yml (hunk 4 above)
Comment (subhamX) [reply docker-compose.yml:40]: I just had to use `stdin_open: true`, and then on my `wsl` it worked flawlessly. It's just that on Linux machines it didn't work. There is no way for me to test if the build is successful for different Linux distros or not. So based on the error output shared by @mathemancer , I changed the script. 

I performed the following four changes based on different solutions on the web.

1. Instead of `stdin_open: true` used `tty: true`.
2. Mount `'/frontend/node_modules'`. [[Link](https://stackoverflow.com/questions/30043872/docker-compose-node-modules-not-present-in-a-volume-after-npm-install-succeeds)]
3. Added `CI=true`
4. Downgraded `react-scripts` to version 3.4.0

With all these changes it's now working fine on @mathemancer PC. I am not sure if removing `CI=true` will have any impact or not. But since there were many people in that thread, who mentioned `CI=true` had some influence on the build status. I thought of leaving it as it is.

File: backend/.env (hunk 11 above)
Comment (subhamX) [reply backend/.env:1]: Yes, it's possible. For dev and prod environment using Docker, it won't be a big problem. 

For dev env without docker, the relative path of that `.env` file will be `../../.env` for the Django application and `build-db-script`. Personally, I didn't like the idea of going too far back like this.

Also, I am not sure if we can move the `.env` file of frontend from the `frontend/` location. `react-scripts` expects the `.env` file in the project root [i.e inside `frontend/`]. There is no mention of giving custom path in the [official docs](https://create-react-app.dev/docs/adding-custom-environment-variables/).

**Comment by subhamX** [comment]
> Just a question, are the .conf files needed?

For `frontend/` yes, for `backend/` no. Since we are no longer using `nginx` for backend we can remove it. For the frontend we still needs a server to host the static files, we need it there. Will remove it very soon from the `backend/`. 🙂

File: docker-compose.yml (hunk 4 above)
Comment (subhamX) [reply docker-compose.yml:40]: @soccerdroid, please advise if I should remove it since this flag does look a bit weird.

File: backend/nginx/nginx.conf (hunk 7 above)
Comment (subhamX) [reply backend/nginx/nginx.conf]: No. It's no longer needed. Just removed them.

Comment (subhamX) [reply backend/nginx/nginx.conf]: No. There are no longer needed. Just removed them.

Comment (subhamX) [reply backend/nginx/nginx.conf]: No. These are no longer needed. Just removed them.

File: backend/Dockerfile.prod (hunk 5 above)
Comment (subhamX) [reply backend/Dockerfile.prod]: Removed these lines.

**Review by mathemancer — APPROVED** [review]
Okay, I think this looks great, good work!

File: backend/.env (hunk 11 above)
Comment (soccerdroid) [reply backend/.env:1]: Ok. We need to point this in the docs then.

**Review by soccerdroid — APPROVED** [review]
